druid/src/widget/textbox.rs

```diff
@@ -310,129 +203,92 @@ impl Widget<String> for TextBox {
-                    edit_action = Some(EditAction::Paste(string));
-                    ctx.request_paint();
+                    self.editor.paste(string, data);
                 }
             }
             Event::KeyDown(key_event) => {
-                let event_handled = match key_event {
+                match key_event {
                     // Tab and shift+tab
-                    k_e if HotKey::new(None, KbKey::Tab).matches(k_e) => {
-                        ctx.focus_next();
-                        true
-                    }
-                    k_e if HotKey::new(SysMods::Shift, KbKey::Tab).matches(k_e) => {
-                        ctx.focus_prev();
-                        true
+                    k_e if HotKey::new(None, KbKey::Tab).matches(k_e) => ctx.focus_next(),
+                    k_e if HotKey::new(SysMods::Shift, KbKey::Tab).matches(k_e) => ctx.focus_prev(),
+                    k_e => {
+                        if let Some(edit) = self.input_handler.handle_event(k_e) {
+                            self.suppress_adjust_hscroll = matches!(edit, EditAction::SelectAll);
+                            self.editor.do_edit(edit, data);
+                            // an explicit request update in case the selection
+                            // state has changed, but the data hasn't.
+                            ctx.request_update();
+                            ctx.request_paint();
+                        }
                     }
-                    k_e if HotKey::new(None, KbKey::Enter).matches(k_e) => {
-                        // 'enter' should do something, maybe?
-                        // but for now we are suppressing it, because we don't want
-                        // newlines.
-                        true
-                    }
-                    _ => false,
                 };
-
-                if !event_handled {
-                    edit_action = BasicTextInput::new().handle_event(key_event);
-                }
-
+                self.reset_cursor_blink(ctx.request_timer(CURSOR_BLINK_DURATION));
                 ctx.request_paint();
             }
             _ => (),
         }
-
-        if let Some(edit_action) = edit_action {
-            let is_select_all = matches!(edit_action, EditAction::SelectAll);
-
-            self.do_edit_action(edit_action, data);
-            self.reset_cursor_blink(ctx.request_timer(CURSOR_BLINK_DURATION));
-
-            if !is_select_all {
-                self.update_hscroll();
-            }
-        }
     }
 
-    fn lifecycle(&mut self, ctx: &mut LifeCycleCtx, event: &LifeCycle, data: &String, _env: &Env) {
+    fn lifecycle(&mut self, ctx: &mut LifeCycleCtx, event: &LifeCycle, data: &T, env: &Env) {
         match event {
             LifeCycle::WidgetAdded => {
                 ctx.register_for_focus();
-                self.text.set_text(data.as_str().into());
+                self.editor.set_text(data.to_owned());
```

Comment:
Now since we have `multiline` support, what if the initial state lensed by a Textbox contains a `\n` char?